tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -1422,8 +1453,16 @@ void SetBranchesHelper(TTree *inputTree, TTree &outputTree, const std::string &i
       // must construct the leaflist for the output branch and create the branch in the output tree
       auto *const leaf = static_cast<TLeaf *>(inputBranch->GetListOfLeaves()->UncheckedAt(0));
       const auto bname = leaf->GetName();
-      const auto counterStr =
-         leaf->GetLeafCount() ? std::string(leaf->GetLeafCount()->GetName()) : std::to_string(leaf->GetLenStatic());
+      auto *sizeLeaf = leaf->GetLeafCount();
+      const auto sizeLeafName = sizeLeaf ? std::string(sizeLeaf->GetName()) : std::to_string(leaf->GetLenStatic());
+      if (sizeLeaf && !outputBranches.Get(sizeLeafName)) { // we need a size branch but it's not there yet
+         const auto sizeTypeStr = TypeName2ROOTTypeName(sizeLeaf->GetTypeName());
+         const auto sizeBufSize = sizeLeaf->GetBranch()->GetBasketSize();
+         // the null address will be reset later when we call SetBranchesHelper for this branch
+         auto *sizeBranch = outputTree.Branch(sizeLeafName.c_str(), (void *)nullptr,
+                                              (sizeLeafName + '/' + sizeTypeStr).c_str(), sizeBufSize);
+         outputBranches.Insert(sizeLeafName, sizeBranch);
+      }
```

Comment:
The code here is a bit tricky to understand without context imho. From reading the linked github issues, I understand the automatic addition of the size leaf can be done only in jitted context, I would add comments about this (possibly in the function documentation).

Reply:
Also, I believe the commit title should also include the fact that this adds branches even when the user doesn't provide them, not only when they are out of order.

Reply:
> // the null address will be reset later when we call SetBranchesHelper for this branch

Does this refer to the fact that if the user provides column names as `{"vec_col", "size_col"}`, the call to `setBranchesHelper` regarding the first `vec_col` will create a placeholder `size_col`, which will only be connected to the real `size_col` in the dataset in a subsequent call to SetBranchesHelper with `size_col` as input?

Reply:
> The code here is a bit tricky to understand without context imho.

I'll expand the comments, thanks.

> From reading the linked github issues, I understand the automatic addition of the size leaf can be done only in jitted context

This does not really have to do with jitted or compiled. Before this PR, if at the point at which we created an output array branch its corresponding size branch was not available, we'd get an error. With this PR, we instead create the size branch on the spot (with a placeholder address that will be properly set later) so we avoid the error _as long as the size branch is part of the set of branches that are being written out_.

> I believe the commit title should also include the fact that this adds branches even when the user doesn't provide them, not only when they are out of order.

That's not the case -- if the user provides an explicit list of output branches and the size branches for some specified array branches are not part of the list, that's an error.

Reply:
ah right, thanks, we can do better in the case of a jitted Snapshot call. I'll add a commit.